### PR TITLE
Std opset export 

### PR DESCRIPTION
**Pull request by neginraoof**
Added export for std (standard deviation) op, plus onnxruntime, caffe2 and expect tests.

### CODE REVIEW

**Comment by neginraoof** [comment]
@houseroad please review

**Comment by houseroad** [comment]
Sure, could you please rebase to master to make sure that CI pass?

**Comment by neginraoof** [comment]
@houseroad I just merged again, but I don't think the failure is related.

**Comment by neginraoof** [comment]
@houseroad Please review

File: torch/onnx/symbolic_opset9.py [hunk 1]
```diff
@@ -1663,6 +1663,40 @@ def gather(g, self, dim, index, sparse_grad=False):
     return g.op("ReduceSum", mul, axes_i=[dim], keepdims_i=0)
 
 
+@parse_args('v', 'is', 'b', 'i')
```
Comment (houseroad) [inline torch/onnx/symbolic_opset9.py]: Do we have b type annotation?

File: torch/onnx/symbolic_opset9.py [hunk 2]
```diff
@@ -1663,6 +1663,40 @@ def gather(g, self, dim, index, sparse_grad=False):
...
+        _unimplemented("std", "Unknown input rank. Cannot compute std along dimensions.")
+
+
+# Since position of optional arguments can change for std, this is a hack to find if first argument
+# is 'dim' or 'unbiased'. As shown below, 'dim' argument could be listed before 'unbiased' :
+# torch.std(input, unbiased=True)
+# torch.std(input, dim, keepdim=False, unbiased=True)
+def std(g, input, *args):
+    if args[0].type().isSubtypeOf(ListType.ofInts()):
+        return _std(g, input, args[0], args[1], args[2])
```
Comment (houseroad) [inline torch/onnx/symbolic_opset9.py]: *arg?

**Review by houseroad — COMMENTED** [review]
Thanks! Btw, we probably should introduce std op in onnx opset 11

File: torch/onnx/symbolic_opset9.py (hunk 1 above)
Comment (neginraoof) [reply torch/onnx/symbolic_opset9.py]: from symbolic_helper, I can see that 'b' is for bool(tval)

File: torch/onnx/symbolic_opset9.py (hunk 2 above)
Comment (neginraoof) [reply torch/onnx/symbolic_opset9.py]: @houseroad what is this about?

File: torch/onnx/symbolic_opset9.py (hunk 1 above)
Comment (houseroad) [reply torch/onnx/symbolic_opset9.py]: Just to confirm, since I was using my phone to do the review on Friday.

File: torch/onnx/symbolic_opset9.py (hunk 2 above)
Comment (houseroad) [reply torch/onnx/symbolic_opset9.py]: using `*` to unpack `args`

**Comment by neginraoof** [comment]
@houseroad review updates please

File: torch/onnx/symbolic_opset9.py [hunk 3]
```diff
@@ -1667,6 +1667,42 @@ def gather(g, self, dim, index, sparse_grad=False):
...
+@parse_args('v', 'is', 'b', 'i')
+def _std(g, input, dim, unbiased, keepdim):
+    if input.type().kind() == "CompleteTensorType" or input.type().kind() == "DimensionedTensorType":
+        sqrd = g.op("Mul", input, input)
+        if dim is None:
+            sqrdmean = g.op("ReduceMean", sqrd, keepdims_i=0)
+            mean = g.op("ReduceMean", input, keepdims_i=0)
+            redudced_dims = input.type().sizes()
+        else:
+            sqrdmean = g.op("ReduceMean", sqrd, axes_i=dim, keepdims_i=keepdim)
```
Comment (houseroad) [inline torch/onnx/symbolic_opset9.py]: also check whether `keepdim` is None or not?

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 4]
```diff
@@ -188,6 +188,24 @@ def forward(self, x):
...
+    def test_std(self):
+        class StandardDeviation(torch.nn.Module):
+            def forward(self, input):
+                return torch.std(input, unbiased=False)
+
+        x = torch.randn(2, 3, 4)
+        model = StandardDeviation()
+        self.run_test(model, x)
+
+    def test_std_along_dims(self):
```
Comment (houseroad) [inline test/onnx/test_pytorch_onnx_onnxruntime.py]: Also test keepdim=True?

File: torch/onnx/symbolic_opset9.py (hunk 3 above)
Comment (neginraoof) [reply torch/onnx/symbolic_opset9.py]: keedim=False by default. Based on the definition of std (def std) below, if dim != None, then keepdim != None.

Comment (neginraoof) [reply torch/onnx/symbolic_opset9.py]: @houseroad

**Comment by houseroad** [comment]
rebase?

**Comment by neginraoof** [comment]
@houseroad build is completed. Could you please review and merge this?

**Comment by neginraoof** [comment]
@bddppq could you please take a look and merge?

**Comment by neginraoof** [comment]
@bddppq could you please take a look and merge?

**Comment by neginraoof** [comment]
@bddppq This PR has been pending for a while. Please review and merge. Thanks

**Comment by neginraoof** [comment]
@bddppq Please review the updates

**Review by bddppq — APPROVED** [review]
